provider/reprovider.go

```diff
@@ -0,0 +1,137 @@
+package provider
+
+import (
+	"context"
+	"gx/ipfs/QmS2aqUZLJp8kF1ihE5rvDGE5LvmKDPnx32w9Z1BW9xLV5/go-ipfs-blockstore"
+	"gx/ipfs/QmZBH87CAPFHcc7cYmBqeSQ98zQ3SX9KUxiYgzPmLWNVKz/go-libp2p-routing"
+	"time"
+)
+
+var (
+	reprovideOutgoingWorkerLimit = 8
+)
+
+type doneFunc func(error)
+
+type Reprovider struct {
+	ctx context.Context
+	queue *Queue
+	tracker *Tracker
+	tick time.Duration
+	blockstore blockstore.Blockstore
+	contentRouting routing.ContentRouting
+	trigger chan doneFunc
+}
+
+// Reprovider periodically re-announces the cids that have been provided. These
+// reprovides can be run on an interval and/or manually. Reprovider also untracks
+// cids that are no longer in the blockstore.
+func NewReprovider(ctx context.Context, queue *Queue, tracker *Tracker, tick time.Duration, blockstore blockstore.Blockstore, contentRouting routing.ContentRouting) *Reprovider {
+	return &Reprovider{
+		ctx: ctx,
+		queue: queue,
+		tracker: tracker,
+		tick: tick,
+		blockstore: blockstore,
+		contentRouting: contentRouting,
+		trigger: make(chan doneFunc),
+	}
+}
+
+// Begin listening for triggers and reprovide whatever is
+// in the reprovider queue.
+func (rp *Reprovider) Run() {
+	go rp.handleTriggers()
+	go rp.handleAnnouncements()
+}
+
+// Add all the cids in the tracker to the reprovide queue
+func (rp *Reprovider) Reprovide() error {
+	cids, err := rp.tracker.Tracking(rp.ctx)
+	if err != nil {
+		return err
+	}
+	for c := range cids {
+		if err := rp.queue.Enqueue(c); err != nil {
```

Comment:
Could we **not** remove entries from the queue after use, and only reset some counters/pointers there (to reduce stress on datastore)?

Reply:
@magik6k can you explain the behavior that you're envisioning and how you think it might impact badger (or I guess another datastore?). I get that we're going to be inserting a lot of tiny records into the datastore and then quickly deleting them. What's the bad effect of doing this? Is it that it fills up the datastore to the configured limit before triggering GC? Or is it that GC becomes way more expensive because badger is slow at collecting lots of little entries?

We could definitely configure a separate datastore as we discussed in the weekly yesterday and we could even call GC on it periodically as part of the providing loop. Would this mitigate the issue?

Reply:
@magik6k I am also curious about the answer to @eingenito's question. After doing research yesterday, I didn't get the sense that we should be overly concerned about this particular usage of badger. Which is probably just me not having all the information you have.

Reply:
I'll start with saying that it's something that definitely needs more testing, and I'll try to get some numbers from go-ds-bench soon(-ish).

Badger keeps data in value logs, separately from LSM tree. I'm not sure about impact of moving lots of data through LSM tree, but my main concern is about the value logs because badger has a rather specific approach to GC. Basically when short-lived data is mixed with it may cause few things:
* Lots of expensive vlog rewrites (I'm not sure how expensive, needs benchmarks)
* Wasted space if short-lived data is mixed with long-lived data in unfortunate proportions (short lived data may never be freed since vlog reclaimable space won't meet one few thresholds)

I'd recommend following the implementation of `RunValueLogGC` and mainly `vlog.doRunGC` - https://github.com/dgraph-io/badger/blob/b9e379e59eac9b828431a6c8f7a553ae38ab88eb/value.go#L1135

Reply:
@magik6k I think I understand your specific concern better. I believe there's a rule in badger that a value log file needs at least 15% of the file to be eligible for removal for it to do anything with that file. So with just the right proportion of long-lived data to short-lived data in a single file, badger will not clean it up. On the flip side, if the proportion is high in the other direction then those files will experience a lot of churn and therefore need to be cleaned up regularly. 

I think, like you said, we need to get data and the cost of these value log rewrites. First, I'm going to finish integrating the rest of the feedback I've received.

Reply:
@Stebalien, @magik6k, @Kubuxu: this work will soon be blocked on getting this issue resolved and so I want to get a few more people involved.

My questions are:
1. Do the potential performance implications, in the grand scheme of things, necessitate me taking a different approach?
2. If we _do not_ have enough data to make this decision right now, does this work get blocked until the data is obtained and the implications assessed?

Just to get everyone on the same page: implementation wise, the `Queue` adds entries to `Datastore` and removes the entry when the work done for that entry is complete. The entries are `cids` and the work is providing them via content routing.

The concern, broadly, is due to the way `Badger` stores its data. There is the potential for often rewriting and processing badger's value log files as data is added/removed. Additionally, in the right proportions, data that is removed may never actually be removed (because it's mixed in the value log file with data that hasn't been removed). But, IMO the primary concern is the former issue, the reprocessing of value log files.

Can we hash this out here?

@magik6k thanks for bringing all of this up and please correct me if this is not the gist of where we are

Reply:
> 2. If we _do not_ have enough data to make this decision right now, does this work get blocked until the data is obtained and the implications assessed?

We could try to work-around this.

I might have missed some bits of prior conversation on this, but what is the reason we have to persist this queue instead of just relying on reprovider? Nodes can shutdown for long periods of time, so old provide records will be expired anyways, forcing us to do full reprovide either way.

Reply:
@magik6k we persist the queue because it can get very big (RAM size big).

Reply:
@Kubuxu thanks for chiming in! (I was out yesterday). @magik6k yes, the reason I was asked to put the queue on disk was due to its being unbounded in RAM (now it's just unbounded on disk).

Reply:
IMO if we need to provide so many CIDs that we can't fit them in reasonable amounts of RAM, we're probably trying to provide way too much.

Assuming 64 bytes per entry, we could fit 16k CIDs per MiB of ram, and while I don't know what were the numbers from simulations @Kubuxu ran in London, I know that 16k provider records is a lot of provider records.

Reply:
Yes, but we don't have probabilistic providers/providers strategies. We used to have unlimited provider queue, then we added backpressure because we were running out of memory.

---

My vote is: let's worry about performance/utilisation of badger when it becomes a problem or when we will be switching to it by default.

Reply:
So, @magik6k's right that we can't really provide more than a on the order of a million records per day and we can easily fit that in memory. However, I also agree with @Kubuxu for the moment: we don't have any fancy provider strategies yet so this queue will grow large (although we _could_ dump the queue and rely on the reprovider when that happens). Really, I have no strong opinions on this.

> so old provide records will be expired anyways, forcing us to do full reprovide either way.

We've discussed avoiding unnecessary full reprovides by remembering our reprovide position/time across restarts.

Reply:
I'm wondering if we can put the provider work into on-disk queues for now and later revisit those queues. The approach taken in this PR does, as I understand it, help with some of the issues we've seen regarding gateways and their unbounded in-memory provider queues. Also, the approach taken in this PR opens the door to implementing more robust and varied providing strategies, which will help greatly with our handling of nodes with lots of content (one of our current ipfs goals).

After this work is in, I think we can start working out important changes to the details while in parallel beginning work on the larger goal we've set for ourselves (probabilistic strategies).

My concern with abandoning the current approach is just that we've spent a lot of time working out the details and worry about it extending even more further than we would like. If the shape of what we've come up with will allow us to move forward with the more impactful changes we are trying to get to, while allowing us to still tweak the details of what we've done so far, then I'm a big vote for moving forward with what we're currently doing.

If we've found a critical flaw, then I obviously wouldn't advocate moving forward. I just want to sort out which things are critical for this work to get merged. I really have no issue going back and rethinking the approach, I just want to get my thoughts out there as I'm conscious of the amount of time we've spent working out what we've got here.

Last note, there are still a few things to do on this PR so I'm not proposing we merge it as is, just that I focus my work on those things that will prevent a merge (again maybe this is one of those thing) while filing issues for things we want to revisit or consider in more depth (but maybe it could be one of these).

Reply:
@michaelavila that sounds like a good approach. If this _does_ cause issues, we can always revert or change our minds later.